enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,993 @@
+   which won't be allowed to be set or deleted. For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.
+   Only 64 items in request and response header list will be allowed.
+   Minimum length of the header value would have to be greater than 0 and less than 4096.
+   Sample fetchers for response dynamic header value allowed are "res.hdr" and "ssl_c_der".
+   Sample fetchers for request dynamic header value allowed are "req.hdr" and "ssl_c_der".
+   Converters allowed for request and response are "lower" and "base64".
+   Actions like Set/Delete for headers can't be set for passthrough routes.
+
+D) route/v1 of openshift/api will contain the validations using kubebuilder tags and CEL for checking Route API fields for Microshift.
+The validations will be same as mentioned in point C).
+
+E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
+The validations will be same as mentioned in point C) except the one for passthrough routes.
+
+F) To validate the header value it has to match a regex.
+
+For the IngressController CR and Route CR (which is used for MicroShift), this validation is defined on the request and response action lists in openshift/api.
+For the Route API as handled by openshift-apiserver (which is used for OpenShift), this is defined in library-go.
+
+This will be the regex:
+```go
+request header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+```go
+response header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+
+Allowed dynamic values for request would be -
+```conf
+%[req.hdr(host),lower]
+%[req.hdr(host),base64]
+%[req.hdr(Host)] if foo
+```
+Allowed dynamic values for response would be -
+```conf
+%[res.hdr(X-Acl-Key),lower]
+%[res.hdr(X-Acl-Key),base64]
+%[req.hdr(Host)] if foo
+```
+Dynamic values allowed for both request and response would be -
+```conf
+%{+Q}[ssl_c_der,base64]
+```
+
+For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
+present in `pkg/route/validation/validation_test.go`
+
+Please note that all values will be quoted before interpolating them into `haproxy.conf`.  Specifically, any single-quote present in the value is escaped, and then the whole value
+is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
+To quote using single-quotes, we will do the following:
+Replace ' with '\''.
+Wrap the entire string in single quotes.
+eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''
```

Comment:
Use ``` `` ``` to quote syntax.  Use a numbered list for steps where order matters.
```suggestion
1. Replace `'` with `'\''`.
2. Wrap the entire string in single quotes.

For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.
```

Reply:
done